AlwaysOnKotlin/compose/src/main/java/com/example/android/wearable/wear/alwayson/AlwaysOnApp.kt

```diff
@@ -0,0 +1,249 @@
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.withContext
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+import java.time.LocalTime
+
+/**
+ * Duration between updates while in active mode.
+ */
+val ACTIVE_INTERVAL: Duration = Duration.ofSeconds(1)
+
+/**
+ * Duration between updates while in ambient mode.
+ */
+val AMBIENT_INTERVAL: Duration = Duration.ofSeconds(10)
+
+const val AMBIENT_UPDATE_ACTION = "com.example.android.wearable.wear.alwayson.action.AMBIENT_UPDATE"
+
+/**
+ * Create a PendingIntent which we'll give to the AlarmManager to send ambient mode updates
+ * on an interval which we've define.
+ */
+private val ambientUpdateIntent = Intent(AMBIENT_UPDATE_ACTION)
+
+@Composable
+fun AlwaysOnApp(
+    ambientState: AmbientState,
+    ambientPing: Long,
+    clock: Clock,
+    activeDispatcher: CoroutineDispatcher
+) {
+    val ambientUpdateAlarmManager = rememberAlarmManager()
+
+    val context = LocalContext.current
+
+    /**
+     * Retrieves a PendingIntent that will perform a broadcast. You could also use getActivity()
+     * to retrieve a PendingIntent that will start a new activity, but be aware that actually
+     * triggers onNewIntent() which causes lifecycle changes (onPause() and onResume()) which
+     * might trigger code to be re-executed more often than you want.
+     *
+     * If you do end up using getActivity(), also make sure you have set activity launchMode to
+     * singleInstance in the manifest.
+     *
+     * Otherwise, it is easy for the AlarmManager launch Intent to open a new activity
+     * every time the Alarm is triggered rather than reusing this Activity.
+     */
+    val ambientUpdatePendingIntent = remember(context) {
+        PendingIntent.getBroadcast(
+            context,
+            0,
+            ambientUpdateIntent,
+            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
+        )
+    }
+
+    /**
+     * A ping used to set up a loopback side-effect loop, to continuously update the time.
+     */
+    var updateDataPing by remember { mutableStateOf(0L) }
+
+    /**
+     * The current instant to display
+     */
+    var currentInstant by remember { mutableStateOf(Instant.now(clock)) }
+
+    /**
+     * The current time to display
+     */
+    var currentTime by remember { mutableStateOf(LocalTime.now(clock)) }
+
+    /**
+     * The number of times the current time and instant have been updated
+     */
+    var drawCount by remember { mutableStateOf(0) }
+
+    fun updateData() {
+        updateDataPing++
+        currentInstant = Instant.now(clock)
+        currentTime = LocalTime.now(clock)
+        drawCount++
+    }
+
+    val lifecycleOwner = LocalLifecycleOwner.current
+
+    /**
+     * Construct a boolean indicating if we are resumed.
+     */
+    val isResumed by produceState(initialValue = false) {
+        lifecycleOwner.repeatOnLifecycle(Lifecycle.State.RESUMED) {
```

Comment:
Interesting, I haven't seen this outside of using it with a coroutine launch. I'll need to try this out. :)

Reply:
I created https://issuetracker.google.com/issues/209684871 in response to this, seems like a decently useful thing to have?

Reply:
Oh cool, I will have a look.